### PR TITLE
svdvals now returns a length-0 sequence of singular values given a size-0 2d array

### PR DESCRIPTION
**Pull request by argriffing**

### CODE REVIEW

**Comment by argriffing** [comment]
> `asarray_validated`?

Fixed.  Also modified `svd` to use this validation.

**Comment by ev-br** [comment]
Why is it desirable to have `svdvals` and `svd` behave differently with zero-size arrays? I'm not a frequent user of either of these, so I might be missing something simple...

```
>>> from scipy.linalg import svd, svdvals
>>> import numpy as np
>>> a = np.zeros((0, 2))
>>> svdvals(a)
array([], dtype=float64)
>>> svd(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/br/virtualenvs/scipy-dev/local/lib/python2.7/site-packages/scipy/linalg/decomp_svd.py", line 97, in svd
    lwork, info = gesdd_lwork(a1.shape[0], a1.shape[1], compute_uv=compute_uv, full_matrices=full_matrices)
ValueError: On entry to DGESDD parameter number 5 had an illegal value 
```

**Comment by argriffing** [comment]
> Why is it desirable to have `svdvals` and `svd` behave differently with zero-size arrays?

Thanks for looking at this!  I agree it's not so great that this PR would de-sync the behavior of `svdvals` and `svd`.

The `svd` implementation could probably be changed.  For example when `compute_uv=0`, `svd` should probably give the same thing as `svdvals`.  But then you will ask why `svd` with the `compute_uv=0` option can deal with empty matrices whereas `svd` with the `compute_uv=1` option cannot.  I haven't though as much about the `U` and `V` matrix outputs of `svd` for empty matrix inputs, but for `svdvals` this is a moot point.  Maybe the `U` and `V` outputs could be appropriately shaped, and maybe they can be identity matrices if orthogonal matrices are required.  There are also complications with left vs. right flags and with full vs. non-full matrix outputs.  All of these additional complications are moot when only `svdvals` is considered and not `svd`.

> I'm not a frequent user of either of these, so I might be missing something simple...

I'll give some background.

There is pretty much a consensus agreement that the empty sum should be 0, the empty product should be 1, and that the concepts of empty lists and 'trivial' matrices with either zero rows or zero columns can be useful.

I would also suggest that empty vector norms and empty matrix norms should be zero, but I'm not 100% certain that this is a convention.  For example in `numpy.linalg`, the only tests for empty vector/matrix norms assert that these norms are zero, but not all norms are included in the tests, possibly for technical reasons (this is complicated by the fact that `np.linalg.norm` can be persuaded to compute functions that do not satisfy any of the inequalities that may characterize norms, for example the minimum absolute value of a vector).  Several matrix norms are defined in terms of singular values, and a slick way to get the 'empty matrix norm == 0' property could be to apply the appropriate function to the empty list of singular values.  For example if a particular matrix norm is defined as the of the sum of the singular values of a matrix, then if the singular values of an empty matrix were defined as an empty list, then this norm of an empty matrix would most naturally be the summation over an empty list, which is zero.

Therefore the natural implementation of empty matrix norms would be one use-case for defining the singular values of an empty matrix to be an empty list.  I don't know of other supporting or conflicting use-cases, and I'm also not sure if there is a mathematical convention.  Suggestions or comments are welcome!

**Comment by ev-br** [comment]
Thanks for the backgound!

I can see why empty sum is zero and empty product is one. As for empty matrices I'm not so sure --- 
I personally only met zero-sized arrays once when the cython code expecting a typed memoryview choked on it :-). But I can imagine it makes sense from some perspective, so take what I'm saying with a large grain of salt.

Basically, I'd think you'll need to special-case an empty matrix somewhere anyway. If the concept of an empty matrix is really useful, and empty matrices are allowed in the linalg routines, then I'd expect `svd(..., compute_uv=True)` with an empty input to return empty matrices for U and V. So that both the sum of the sigular values if zero, their product is one, and the norm of both U and V is zero. Oh, no, golly, then these matrices are not orthogonal. But if they are orthogonal, they cannot be empty, can they? Or if you return identity matrices for U and V, and an empty one for the singular values, S, you cannot maintain that if S, U and V are the SVD decomp of A, then A = U @ S @ V.T ? 

Basically, my gut feeling is that 
* `svdvals` being out of sync with `svd` is weird.
* is there really a reason to keep `svdvals` as a separate function instead of recommending `svd(.., compute_uv=False)`?
* Either linalg operations can be defined consistently for the empty matrices and then they should be allowed everywhere; otherwise they should not be allowed anywhere, and special-casing for the empty sum and empty product should be pushed up to the user level. 

Ok, I'm hiding back behind a spoonful of salt to accompany whatever I said above.

**Comment by argriffing** [comment]
That sounds pretty reasonable.  Maybe this PR should be replaced by a PR that gives more informative error messages for size-0 inputs for `svdvals` and `svd`, or even adds non-emptiness to the list of pre-conditions before calling lower level functions (this currently filters out sparse matrices, masked arrays, and object arrays)?

**Comment by ev-br** [comment]
(Just noticed a possibly related numpy issue, https://github.com/numpy/numpy/issues/5419).

Don't let my gut feeling block progress :-). Maybe the right way is to proceed is to merge your changes for svdvals, and clearly document that svdvals only differs from svd(..., compoute_uv=True) in the handling of empty input --- how does this sound?
This only requires a one-line addition to the docstring of svdvals, it seems...

**Comment by ev-br** [comment]
For example, here is a shot at phrasing it: 
https://github.com/ev-br/scipy/commit/fc4a082ff91f9ca6807a2939aa3423dacf89869b

If you either take over this commit or come up with a better wording, I'm +1 for merging this PR.

Just to be clear: I'm changing my mind on the consistency of svdvals and svd (insert an obligatory Knuth quote here). Since `svdvals` exists, it makes sense to have it doing things which make sense for singular values but not for the orthogonal matrices from the full SVD decomposition.

**Comment by ev-br** [comment]
Bump. @argriffing Maybe we should finish this for 0.16, what do you think?

**Comment by argriffing** [comment]
@ev-br I've `git cherry-pick evbr/pr/4365`'d the commit

**Comment by ev-br** [comment]
OK, it's been open for long enough and did not attract any further comments, so let's put it in. Thanks Alex and sorry for slowing this down.
